usr/sbin/p11sak/p11sak.c

```diff
@@ -14,3457 +14,1235 @@
+#ifdef NID_modp_1536
+    { .value = "modp1536", .args = NULL,
+      .private = { .num = NID_modp_1536 }, },
+#endif
+#ifdef NID_modp_2048
+    { .value = "modp2048", .args = NULL,
+      .private = { .num = NID_modp_2048 }, },
+#endif
+#ifdef NID_modp_3072
+    { .value = "modp3072", .args = NULL,
+      .private = { .num = NID_modp_3072 }, },
+#endif
+#ifdef NID_modp_4096
+    { .value = "modp4096", .args = NULL,
+      .private = { .num = NID_modp_4096 }, },
+#endif
+#ifdef NID_modp_6144
+    { .value = "modp6144", .args = NULL,
+      .private = { .num = NID_modp_6144 }, },
+#endif
+#ifdef NID_modp_8192
+    { .value = "modp8192", .args = NULL,
+      .private = { .num = NID_modp_8192 }, },
+#endif
+    { .value = "DH-PARAM-PEM-FILE", .args = NULL,
+      .private = { .num = 0 },
+      .any_value = &opt_pem_file, },
+    { .value = NULL, },
+};
+
+static const struct p11sak_arg p11sak_generate_dh_args[] = {
+    { .name = "GROUP", .type = ARG_TYPE_ENUM, .required = true,
+      .enum_values = p11sak_dh_group,
+      .value.enum_value = &opt_group,
+      .description = "The Diffie-Hellman FFC group name "
+                     "or the name of a DH parameters PEM file:", },
+    { .name = "PRIV-BITS", .type = ARG_TYPE_NUMBER, .required = false,
+      .value.number = &opt_keybits_num,
+      .description = "Size of the DH private key in bits (optional).",},
+    { .name = NULL, },
+};
+
+static const struct p11sak_arg p11sak_generate_dsa_args[] = {
+    { .name = "DSA-PARAM-PEM-FILE", .type = ARG_TYPE_STRING, .required = true,
+      .value.string = &opt_pem_file,
+      .description = "The name of a DSA parameters PEM file.", },
+    { .name = NULL, },
+};
+
+#define DECLARE_CURVE_INFO(name, size)                                         \
+    static const struct curve_info name ## _info = {                           \
+        .oid = name, .oid_len = sizeof(name), .bitsize = size,                 \
+    }
+
+#define DECLARE_CURVE_VALUE(name)                                              \
+    { .value = # name, .args = NULL, .private = { .ptr = &name ## _info, }, }
+
+DECLARE_CURVE_INFO(prime256v1, 256);
+DECLARE_CURVE_INFO(prime192v1, 192);
+DECLARE_CURVE_INFO(secp224r1, 224);
+DECLARE_CURVE_INFO(secp384r1, 384);
+DECLARE_CURVE_INFO(secp521r1, 521);
+DECLARE_CURVE_INFO(secp256k1, 256);
+DECLARE_CURVE_INFO(brainpoolP160r1, 160);
+DECLARE_CURVE_INFO(brainpoolP160t1, 160);
+DECLARE_CURVE_INFO(brainpoolP192r1, 192);
+DECLARE_CURVE_INFO(brainpoolP192t1, 192);
+DECLARE_CURVE_INFO(brainpoolP224r1, 224);
+DECLARE_CURVE_INFO(brainpoolP224t1, 224);
+DECLARE_CURVE_INFO(brainpoolP256r1, 256);
+DECLARE_CURVE_INFO(brainpoolP256t1, 256);
+DECLARE_CURVE_INFO(brainpoolP320r1, 320);
+DECLARE_CURVE_INFO(brainpoolP320t1, 320);
+DECLARE_CURVE_INFO(brainpoolP384r1, 384);
+DECLARE_CURVE_INFO(brainpoolP384t1, 384);
+DECLARE_CURVE_INFO(brainpoolP512r1, 512);
+DECLARE_CURVE_INFO(brainpoolP512t1, 512);
+DECLARE_CURVE_INFO(curve25519, 256);
```

Comment:
The PoP says: "For Ed25519 the most significant bit of the most significant byte is not needed for 255 bit format though the standard requires this bit to be a zero." See description of KDSA instruction. In libzpc, we in fact are using a bitlength of 255.